sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledSizedHashJoinExec.scala

```diff
@@ -826,35 +835,44 @@ object GpuShuffledAsymmetricHashJoinExec {
       val exprs = BoundJoinExprs.bind(joinType, leftKeys, leftOutput, rightKeys,
         rightOutput, condition, buildSide)
       val buildQueue = mutable.Queue.empty[T]
-      val (buildRows, buildSize) = closeOnExcept(buildQueue) { _ =>
+      val (mayTruncatedBuildRows, mayTruncatedBuildSize) = closeOnExcept(buildQueue) { _ =>
         fetchProbeTargetSize(probeBuildIter, buildQueue, gpuBatchSizeBytes)
       }
       val baseBuildIter = setupForJoin(buildQueue, rawBuildIter, exprs.buildTypes,
         gpuBatchSizeBytes, metrics)
-      if (buildRows <= Int.MaxValue && buildSize <= gpuBatchSizeBytes) {
+      if (mayTruncatedBuildRows <= Int.MaxValue && mayTruncatedBuildSize <= gpuBatchSizeBytes) {
         assert(!probeBuildIter.hasNext, "build side not exhausted")
         getJoinInfoSmallBuildSide(joinType, buildSide, condition, exprs,
-          baseBuildIter, buildRows, buildSize,
+          baseBuildIter, mayTruncatedBuildRows, mayTruncatedBuildSize,
           rawStreamIter, gpuBatchSizeBytes, metrics)
       } else {
         // The natural build side does not fit in a single batch, so use the stream side
         // as the hash table if we can fit it in a single batch.
         val streamQueue = mutable.Queue.empty[T]
-        val (streamRows, streamSize) = closeOnExcept(streamQueue) { _ =>
+        val (mayTruncatedStreamRows, mayTruncatedStreamSize) = closeOnExcept(streamQueue) { _ =>
           fetchProbeTargetSize(probeStreamIter, streamQueue, gpuBatchSizeBytes)
         }
         val streamIter = setupForJoin(streamQueue, rawStreamIter, exprs.streamTypes,
           gpuBatchSizeBytes, metrics)
-        if (streamRows <= Int.MaxValue && streamSize <= gpuBatchSizeBytes) {
+        if (mayTruncatedStreamRows <= Int.MaxValue && mayTruncatedStreamSize <= gpuBatchSizeBytes) {
           assert(!probeStreamIter.hasNext, "stream side not exhausted")
-          metrics(BUILD_DATA_SIZE).set(streamSize)
+          metrics(BUILD_DATA_SIZE).set(mayTruncatedStreamSize)
           val flippedSide = flipped(buildSide)
-          JoinInfo(joinType, flippedSide, streamIter, streamSize, None, baseBuildIter,
+          JoinInfo(joinType, flippedSide, streamIter, mayTruncatedStreamSize, None, baseBuildIter,
             exprs.flipped(joinType, flippedSide, condition))
         } else {
+          // Here buildSize provided to JoinInfo is important for deciding the partition number.
+          // We should provide the actual buildSize instead of the truncated one.
+          // By calling fetchProbeTargetSize again, we'll move all batches to the queue, and
+          // at the same time we'll get the actual buildSize.
+          val (_, remainingBytes) = closeOnExcept(buildQueue) { _ =>
```

Comment:
I am not a big fun of this solution which moves all the bathes to the queue after a call to `setupForJoin`, since it will implicitly change the queue while `setupForJoin` can not sense. Then this forces a limit on `setupForJoin` that it must refer to the input queue directly. 
Shall we have a follow-up issue to improve this ?

Reply:
hi @firestarman , the reason why I'm not draining the iterator after setupForJoin is because the iterator returned by setupForJoin is unspillable, whereras what we have in the queue is still spillable. The current implementation, as you   pointed out, might be a little bit tricky and imposes some contraints on how we should use the `queue`, but it can save the cost of registering the batch again. I can add some comments to remind people to be careful about the `queue`, what you think?

Reply:
File https://github.com/NVIDIA/spark-rapids/issues/12355 for this.

Reply:
Note that the queue here is not spillable (`HostHostAsymmetricJoinSizer`). It is host memory, afaik, but still adds another place where we need to go back later and try to control the memory usage.

Reply:
For kudo case, the elements in the queue is moslty KudoSerializedTableColumn, which has already changed to spillable after https://github.com/NVIDIA/spark-rapids/issues/12215 @abellina . But for non-kudo case you're right. Pls correct me if I'm wrong

Reply:
Ok I think we need to resolve that these are not spillable for the host. Kudo isn't going to be on by default in 25.04, and this change will be on by default for all users. 

I think we are going to have to solve the issue of `queue` not holding spillable elements. We will need to inspect each batch as we pop them from the iterator, and decide: is it host batch? then we need it to be a `SpillableHostColumnarBatch`, is it a gpu batch? then we need to hold a `SpillableColumnarBatch`. If it is Kudo, then we do that differently, but it should be consistent, all ColumnarBatch instances in that case would be Kudo. 

If we don't solve the above, then we run the risk of running out of memory on the host where we didn't use to before.